Add documentation - Main Concepts - Querying (#1863)
kuzzleio/kuzzle: doc/2/guides/main-concepts/3-querying/index.md

::: info
By default, the [document:search](/core/2/api/controllers/document/search) action returns only 10 documents.  
The number of returned documents can be changed with the `size` option.  
There is a limit of 10000 documents that can be retrieved with a request and the limit can be configured under the `limits.documentsFetchCount` [configuration](/core/2/guides/advanced/advanced/8-configuration) key.
:::

Those methods are explained in the next sections and they are already implemented in our SDKs in the `SearchResult` class. (e.g. [SearchResult.next](/sdk/js/7/core-classes/search-result/next) method in the Javascript SDK)

All the methods explained here are available by default in our SDKs through the `SearchResult` class. This class allows to navigate through your paginated results with ease by calling the `SearchResult.next` method.

**Example:** _Paginate search result using the scroll method_
:::: tabs

::: Javascript

```js
let result = await sdk.document.search('ktm-open-data', 'thamel-taxi', {
  query: {
    term: { city: 'Antalya' }
  }
},
{ scroll: '5s' });

while (result) {
  console.log(result.hits);
  result = await result.next();
}
```

:::
::: Dart

```dart

> var result = await kuzzle

Author: @jenow @Yoann-Abbes is it correct?

Reviewer: I added corrections
You can see an example of search with scroll here
https://docs.kuzzle.io/sdk/dart/2/core-classes/search-result/next/

  .document
  .search('ktm-open-data', 'thamel-taxi', query: {
    'query': {
      'term': {
        'city': 'Antalya'
      }
    }

>   },
>   { 'scroll': '5s' });

Reviewer: ```suggestion
  },
  'scroll': '5s');
```

while (result) {
  print(result.hits);
  result = await result.next();
}
```

:::

::: Kotlin

```kotlin

> val term: ConcurrentHashMap<String, Any?> =

Author: @jenow @Yoann-Abbes is this example correct?
Is there any way to not declare 3 intermediate object for the search query?

  ConcurrentHashMap<String, Any?>().apply {
    put("city", "Antalya")
  }
val query: ConcurrentHashMap<String, Any?> =
  ConcurrentHashMap<String, Any?>().apply {
    put("term", term)
  }
val searchQuery: ConcurrentHashMap<String, Any?> =
  ConcurrentHashMap<String, Any?>().apply {
    put("query", query)
  }

> val options: ConcurrentHashMap<String, Any?> =
>   ConcurrentHashMap<String, Any?>().apply {
>     put("scroll", "5s)
>   }
> 
> val results = kuzzle
>   .documentController
>   .search("ktm-open-data", "thamel-taxi", searchQuery, options).get();
> 

Reviewer: ```suggestion

val results = kuzzle
  .documentController
  .search("ktm-open-data", "thamel-taxi", searchQuery, "1s").get();

```

We do not use `options` object in this SDK

Author: But if I want to pass the size and from option without `scroll` ?

Reviewer: Just do it :)
https://docs.kuzzle.io/sdk/jvm/1/core-classes/search-result/next/

while (result) {
  println(result.hits);

  result = result.next().get();
}
```

:::

::: Csharp

```csharp

> SearchOptions options = new SearchOptions();

Author: @scottinet  is it correct?

# Querying

Kuzzle directly exposes [Elasticsearch's query language](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl.html) in a secure way. 

It is possible for a client to send requests to **retrieve documents from any authorized collection**.

Search query can be passed in the body of the [document:search](/core/2/api/controllers/document/search) action, then will be forwarded to Elasticsearch [Search API](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/search-search.html) endpoint.

Elasticsearch supports many keyword in a search query root level. For security reasons Kuzzle only supports the following keywords:
  - `aggregations`
  - `aggs`
  - `collapse`
  - `explain`
  - `from`
  - `highlight`
  - `query`
  - `search_after`
  - `search_timeout`
  - `size`
  - `sort`
  - `_name`
  - `_source`
  - `_source_excludes`
  - `_source_includes`

::: info

> Any other keyword will be ignored by Kuzzle and not followed to Elasticsearch.

Author: Ahh yes true :facepalm:

    }
  ]
}'
```

</details>

### `term` clause

The [term](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl-term-query.html) clause allows to returns documents that **contain an exact value** in a provided field.

Term clause should be used on fields with the [keyword](/core/2/guides/main-concepts/2-data-storage#mappings-properties) type.

::: info
You can use the `term` clause to find documents based on a precise value such as a price, a product ID, or a username.
:::

**Example:** _Search for documents containing an exact field value_ 
```bash
kourou document:search ktm-open-data thamel-taxi '{
  term: { name: "Jenow" }
}'
```

::: info
The search query content will be injected in the request body inside the `query` property:
```js
{
  query: {
    term: { name: "Jenow" }
  }
}
```
:::

### `match` clause

The [match](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl-match-query.html) clause allows to returns documents **that match (could be approximately)** a provided `text`, `number`, `date` or `boolean` field.

The match query is the standard query for **performing a full-text search**, including options for fuzzy matching.

::: info

> The `match` clause to find documents with fields containing a value. The `match` clause as well as the content of a `text` fields are [analyzed](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/analysis-analyzers.html) by Elasticsearch before performing the query.

Author: I wanted to say that the Elasticsearch analyze the `match` clause value. For example it will separate the words, trim spaces, lowercase everything etc

That what they say in the documentation https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl-match-query.html#query-dsl-match-query

> The `match` clause to find documents with fields containing a value. The `match` clause as well as the content of a `text` fields are [analyzed](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/analysis-analyzers.html) by Elasticsearch before performing the query.

Reviewer 2: Well, then I think that the first sentence is confusing and that it doesn't bring value to the explanation. I think that this note is a lot better without it. wdyt?

```suggestion
The `match` clause as well as the content of a `text` fields are [analyzed](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/analysis-analyzers.html) by Elasticsearch before performing the query.
```

> The `match` clause to find documents with fields containing a value. The `match` clause as well as the content of a `text` fields are [analyzed](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/analysis-analyzers.html) by Elasticsearch before performing the query.

Author: Yes I agree too it's better. First I wanted to explain that it's not an exact match but I think the example did it better

  { "_id": "asc" }
]'
```

## Pagination

[Several pagination methods](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html) are available using Elasticsearch and Kuzzle.

They allow to find all documents matching a search query.

::: info
By default, the [document:search](/core/2/api/controllers/document/search) action returns only 10 documents.  
The number of returned documents can be changed with the `size` option.  
There is a limit of 10000 documents that can be retrieved with a request and the limit can be configured under the `limits.documentsFetchCount` [configuration](/core/2/guides/advanced/advanced/8-configuration) key.
:::

Those methods are explained in the next sections and they are already implemented in our SDKs in the `SearchResult` class. (e.g. [SearchResult.next](/sdk/js/7/core-classes/search-result/next) method in the Javascript SDK)

All the methods explained here are available by default in our SDKs through the `SearchResult` class. This class allows to navigate through your paginated results with ease by calling the `SearchResult.next` method.

**Example:** _Paginate search result using the scroll method_

> :::: tabs

Reviewer 3: Tabs rendering looks broken in the preview 
<img width="824" alt="image" src="https://user-images.githubusercontent.com/7868838/98931122-e7c8c900-24dd-11eb-9812-e65b399ceed1.png">

:::

**Example:** _Use sort and size to navigate through pages of results_

```bash
kourou sdk:query document:search \
  -i ktm-open-data -c thamel-taxi -a size=2 --body '{
    sort: [
      { _id: "desc" }
    ],
  }'

# {
#   "hits": [
#     {
#       "_id": "liaa",
#         # ...
#       }
#     },
#     {
#       "_id": "jenow",
#         # ...
#       }
#     }
#   ],
#   "total": 4
# }
```

Then we will include the `_id` of the last document in the `search_after` parameter:
```bash
kourou sdk:query document:search \
  -i ktm-open-data -c thamel-taxi -a size=2 --body '{
    sort: [
      { _id: "desc" }
    ],
    search_after: ["jenow"]
  }'

# {
#   "hits": [
#     {
#       "_id": "domisol",
#         # ...
#       }
#     },
#     {
#       "_id": "aschen",
#         # ...
#       }
#     }
#   ],
#   "total": 4
# }
```

::: info
Because this method does not freeze the search results between two calls, **there can be missing or duplicated documents between two result pages**.  
This method efficiently mitigates the costs of scroll searches, but returns less consistent results: it's a middle ground, **ideal for real-time search requests**.  

> Also it's not possible to retrieve more than 10000 documents with this method.

Reviewer 2: False: this method is capable of returning as many documents as needed (given that each page is less than 10k documents).

```suggestion
```

:::

### Paginate with Scroll Cursor

> The `scroll` parameter can be specified in the search query to allows the usage of the [document:scroll](/core/2/api/controllers/document/scroll) action to **use a cursors to move through paginated results**.

Reviewer 2: ```suggestion
The `scroll` parameter can be specified in the search query to allow the usage of the [document:scroll](/core/2/api/controllers/document/scroll) action. This option creates **a forward-only cursor** to move through paginated results.
```

:::

The search action will return a `scrollId` that you have to use with the [document:scroll](/core/2/api/controllers/document/scroll) to get the next page of results.

**Example:** _Paginate search with scroll_
```bash
kourou sdk:query document:search \
  -i ktm-open-data -c thamel-taxi -a scroll=30s -a size=2

# {
#   "hits": [
#     {
#       "_id": "aschen",
#         # ...
#       }
#     },
#     {
#       "_id": "jenow",
#         # ...
#       }
#     }
#   ],
#   "remaining": 2,
#   "scrollId": "DXF1ZXJ5QW5kRmV0Y2gBAAAAAAAAACUWblZVaDV4UnNTck9mLU9wczZUVlBkUQ==",
#   "total": 4
# }
```

Then we have to pass the `scrollId` parameter to the [document:scroll](/core/2/api/controllers/document/scroll) action:
```bash
kourou sdk:query document:scroll -a scrollId=<scroll-id>

# {
#   "hits": [
#     {
#       "_id": "liia",
#         # ...
#       }
#     },
#     {
#       "_id": "domisol",
#         # ...
#       }
#     }
#   ],
#   "remaining": 0,
#   "scrollId": "DXF1ZXJ5QW5kRmV0Y2gBAAAAAAAAACUWblZVaDV4UnNTck9mLU9wczZUVlBkUQ==",
#   "total": 4
# }
```

::: warning

> When using a cursor with the `scroll` option, Elasticsearch has to duplicate the transaction log to keep the same result during the entire scroll session.

Reviewer 2: ```suggestion
When using a cursor with the `scroll` option, Elasticsearch has to duplicate the transaction log to keep consistent results during the entire scroll session.
```

> It **can lead to memory leaks** if a scroll duration too great is provided, or if too many scroll sessions are open simultaneously.

Reviewer 2: ```suggestion
It **can lead to memory leaks** if a scroll duration too great is provided, or if too many scroll sessions are open simultaneously.  

By default, Kuzzle sets a maximum scroll duration of 1 minute. This can be changed in the kuzzlerc configuration file.
```

:::

> 

Reviewer 2: I would add an additional information there: 

```suggestion

::: info
Kuzzle automatically destroys scroll cursors that have been consumed. Invoking the `document:scroll` action on a scroll ID whose results have been completely fetched will lead to an "unknown scroll ID" error.
:::
```

## Aggregations

Elasticsearch allows data to be **grouped together** to **build complex summaries** of the data through a mechanism called aggregations.

This mechanism allows to **structure the data returned in the response** in order to easily build dashboards, graphs, maps, etc.

> More informations on [Elasticsearch Aggregations](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/search-aggregations.html)

Reviewer 2: ```suggestion
More information on [Elasticsearch Aggregations](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/search-aggregations.html)
```

The search query content will be injected in the request body inside the `query` property:
```js
{
  query: {
    term: { name: "Jenow" }
  }
}
```
:::

### `match` clause

The [match](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl-match-query.html) clause allows to returns documents **that approximately match** a provided `text`, `number`, `date` or `boolean` field.

The match query is the standard query for **performing a full-text search**. As thus, it includes options for fuzzy matching.

::: info
The `match` clause to find documents with fields containing a value. The `match` clause as well as the content of a `text` fields are [analyzed](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/analysis-analyzers.html) by Elasticsearch before performing the query.
:::

**Example:** _Search for documents roughly matching the provided field value_ 
```bash
kourou document:search ktm-open-data thamel-taxi '{
  match: { description: "java" }
}'
```

### `range` clause

The [range](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl-range-query.html) clause allows to return documents **that contain value within a provided range**.

> It can be used with `number` and `date` fields, but also with `keyword` ones, even if it's less common.

Author: Actually that's what we do with the query generator in the admin console :sweat_smile: 
But yes you're right that's definitely not common or a good practice :+1:

**Example:** _Search for documents with their ids_ 
```bash
kourou document:search ktm-open-data thamel-taxi '{
  ids: {
    values: ["aschen", "liia"]
  }
}'
```

::: info
If you only have an `ids` clause in your search query then you should rather use the [document:mGet](/core/2/api/controllers/document/m-get) action.
:::

## Boolean Query

It is possible to **combine several clauses in the same query** by using a [Boolean Query](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl-bool-query.html).

The following 4 operands are available:
 - `must`: The clause **must appear** in matching documents and will contribute to the score.
 - `filter`: The clause **must appear** in matching documents. The score of the query will be ignored.
 - `should`: The clause **should appear** in the matching document.
 - `must_not`: The clause **must not appear** in the matching documents. The score of the query will be ignored.

**Example:** _Combining clauses to create an "AND" like search query_
```bash
kourou document:search ktm-open-data thamel-taxi '{
  bool: {
    filter: [
      { term: { city: "Tirana" } },
      { range: { age: { gte: 30 } } }
    ]
  }
}'
```

> **Example:** _Combining clauses to create an "OR" like search query_

Reviewer 4: ```suggestion
**Example:** _Combining clauses to create an "OR"-like search query_
```

```bash
kourou document:search ktm-open-data thamel-taxi '{
  bool: {
    should: [
      { term: { city: "Siccieu" } },
      { term: { city: "Kathmandu" } },
    ]
  }
}'
```

## Sorting

> Elasticsearch offers the possibility to **[sort the results](https://www.elastic.co/guide/en/elasticsearch/reference/current/sort-search-results.html) by one or more fields**.

Reviewer 4: ```suggestion
Elasticsearch enables to **[sort the results](https://www.elastic.co/guide/en/elasticsearch/reference/current/sort-search-results.html) by one or more fields**.
```

**Example:** _Sort by multiple fields_
```bash
kourou document:search ktm-open-data thamel-taxi --sort '[
  { city: "asc" },
  { age: "desc" }
]'
```

A very common sort with Kuzzle is to use [Kuzzle Metadata](/core/2/guides/main-concepts/2-data-storage#kuzzle-metadata) to sort by creation date:

```bash
kourou document:search ktm-open-data thamel-taxi --sort '[
  { "_kuzzle_info.createdAt": "asc" }
]'
```

Finally, if you want to always ensure the same order for your results you can sort on the `_id` field:

```bash
kourou document:search ktm-open-data thamel-taxi --sort '[
  { "_id": "asc" }
]'
```

## Pagination

[Several pagination methods](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html) are available using Elasticsearch and Kuzzle.

They allow to find all documents matching a search query.

::: info
By default, the [document:search](/core/2/api/controllers/document/search) action returns only 10 documents.  
The number of returned documents can be changed with the `size` option.  

> There is a limit of 10000 documents that can be retrieved with a single search request, without pagination.

Reviewer 4: ```suggestion
Elasticsearch does not allow a `search` request to return more than 10000 documents, no matter what pagination parameters are set. Use the `scroll` parameter and the [document:scroll](/core/2/api/controllers/document/scroll) action to paginate more results.
```

Author: I updated the comment, tell me what do you think

Reviewer 4: LGTM

:::

> Those methods are explained in the next sections and they are already implemented in our SDKs in the `SearchResult` class. (e.g. [SearchResult.next](/sdk/js/7/core-classes/search-result/next) method in the Javascript SDK)

Reviewer 4: I'd delete this and keep the next one.
```suggestion
```

> All the methods explained here are available by default in our SDKs through the `SearchResult` class. This class allows to navigate through your paginated results with ease by calling the `SearchResult.next` method.

Reviewer 4: ```suggestion
All the methods explained here are available by default in our SDKs through the `SearchResult` class. This class allows to navigate through your paginated results with ease by calling the [SearchResult.next](/sdk/js/7/core-classes/search-result/next) method.
```

Author: I didn't wanted to put a link here since the `SearchResult.next` method is available in all SDKs

**Example:** _Paginate search results using the scroll method_

> :::: tabs

Reviewer 4: ```suggestion
::: tabs
```

Author: We actually need 3 colon to start the tabs area

**Example:** _Paginate search with scroll_
```bash
kourou sdk:query document:search \
  -i ktm-open-data -c thamel-taxi -a scroll=30s -a size=2

# {
#   "hits": [
#     {
#       "_id": "aschen",
#         # ...
#       }
#     },
#     {
#       "_id": "jenow",
#         # ...
#       }
#     }
#   ],
#   "remaining": 2,
#   "scrollId": "DXF1ZXJ5QW5kRmV0Y2gBAAAAAAAAACUWblZVaDV4UnNTck9mLU9wczZUVlBkUQ==",
#   "total": 4
# }
```

Then we have to pass the `scrollId` parameter to the [document:scroll](/core/2/api/controllers/document/scroll) action:
```bash
kourou sdk:query document:scroll -a scrollId=<scroll-id>

# {
#   "hits": [
#     {
#       "_id": "liia",
#         # ...
#       }
#     },
#     {
#       "_id": "domisol",
#         # ...
#       }
#     }
#   ],
#   "remaining": 0,
#   "scrollId": "DXF1ZXJ5QW5kRmV0Y2gBAAAAAAAAACUWblZVaDV4UnNTck9mLU9wczZUVlBkUQ==",
#   "total": 4
# }
```

::: warning
When using a cursor with the `scroll` option, Elasticsearch has to duplicate the transaction log to keep consistent results during the entire scroll session.

> It **can lead to memory leaks** if a scroll duration too great is provided, or if too many scroll sessions are open simultaneously.  

Reviewer 4: ```suggestion
It **can lead to memory issues** if a scroll duration too high is provided, or if too many scroll sessions are open simultaneously.  
```

> By default, Kuzzle sets a maximum scroll duration of 1 minute. This can be changed in the kuzzlerc configuration file under the key `services.storageEngine.maxScrollDuration`.

Reviewer 4: ```suggestion
By default, Kuzzle sets a maximum scroll duration of 1 minute. This can be changed in the `.kuzzlerc` configuration file under the key `services.storageEngine.maxScrollDuration`.
```
